flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java

```diff
@@ -0,0 +1,228 @@
+
+        @Override
+        public void onException(Exception exception) {
+            handlers.forEach((k, h) -> h.handleExceptionEvent(exception));
+        }
+
+        private Watch watch(String name, WatchCallbackHandler<R> watch) {
+            final String resourceKey = getResourceKey(name);
+            final String watchId = UUID.randomUUID().toString();
+            final CompletableFuture<Void> closeFuture = new CompletableFuture<>();
+            executorService.submit(
+                    () -> {
+                        final EventHandler eventHandler =
+                                handlers.computeIfAbsent(
+                                        resourceKey, key -> new EventHandler(resourceKey));
+                        eventHandler.addWatch(watchId, watch);
+                    });
+            closeFuture.whenCompleteAsync(
+                    (ignored, error) -> {
+                        if (error != null) {
+                            log.error("Unhandled error while closing watcher.", error);
+                        }
+                        final boolean removeHandler =
+                                handlers.get(resourceKey).removeWatch(watchId);
+                        if (removeHandler) {
+                            handlers.remove(resourceKey);
+                        }
+                    },
+                    executorService);
+            return () -> closeFuture.complete(null);
+        }
+
+        private void onResourceEvent(T obj) {
+            executorService.submit(
+                    () -> findHandler(obj).ifPresent(EventHandler::handleResourceEvent));
+        }
+
+        private Optional<EventHandler> findHandler(T obj) {
+            final String resourceKey = getResourceKey(obj.getMetadata().getName());
+            return Optional.ofNullable(handlers.get(resourceKey));
+        }
+    }
+
+    private class EventHandler {
+        private final String resourceKey;
+        private final Map<String, WatchCallbackHandler<R>> handlers = new HashMap<>();
+
+        private T currentResource;
+
+        private EventHandler(String resourceKey) {
+            this.resourceKey = resourceKey;
+        }
+
+        private void addWatch(String id, WatchCallbackHandler<R> handler) {
+            log.info("Starting to watch for {}, watching id:{}", resourceKey, id);
+            handlers.put(id, handler);
+            if (currentResource != null) {
+                final List<R> events =
+                        Collections.singletonList(eventWrapper.apply(currentResource));
+                handler.onAdded(events);
+            }
+        }
+
+        private boolean removeWatch(String id) {
+            handlers.remove(id);
+            log.info("Stopped to watch for {}, watching id:{}", resourceKey, id);
+            return handlers.isEmpty();
+        }
+
+        private void handleResourceEvent() {
+            T newResource = sharedIndexInformer.getIndexer().getByKey(resourceKey);
```

Comment:
what is the reason for "diffing" events manually? Isn't the informer doing the same thing already?

Reply:
i think you are right, we can ensure the sequentiality without "diffing" events manually in a single thread.

Reply:
thanks

Reply:
@dmvk seems it is still needed to do the 'diffing' locally, the `EventHandler` need to get the base state from indexer.  So it can obtained a newer state from `EventHandler` which has not arrived the `EventHandler`. So use the event directly may lead to conflict.

Reply:
That sounds weird. Do you have any test case that's failing with this approach?

I don't think it's worth blocking a PR on this, but I'd give a quick look to get a better understanding of how the informer interface behaves.